drivers/misc/ledpwm.c

```diff
@@ -8,23 +8,50 @@
 #include <linux/module.h>
 #include <linux/kernel.h>
 #include <linux/init.h>
+#include <linux/errno.h>
 #include <linux/ioport.h>
 #include <linux/io.h>
 #include <linux/jiffies.h>
 #include <linux/timer.h>
 
+#include <linux/fs.h>
+#include <linux/cdev.h>
+#include <linux/uaccess.h>
+#include <linux/device.h>
+#include <linux/delay.h>
+
+// -----------------------------------------------------------------
+// PWM channel definitions
+static size_t __iomem *led_regs; // module intern pointer to the memory region
+
 #define PWM_REG_BASE 0xFF203080UL
-#define PWM_NUM_CHANNELS 10
+#define PWM_NUM_CHANNELS 8 // only map 8 led channels to the
 #define PWM_REG_SIZE (PWM_NUM_CHANNELS * sizeof(uint32_t))
 
 #define PWM_MAX 0x7FFU
 
+// -----------------------------------------------------------------
+// Constants for the timer/running led
 #define NUM_LEDS 8
 #define STARTUP_DELAY 3000
 #define TIMER_DELAY 125
 
-static void __iomem *led_regs;
+#define MULTIPLE_WRITE_DELAY 200
+
+static int set_led(size_t __iomem *reg_base, int ledNr, uint32_t value);
+
+// -----------------------------------------------------------------
+// PWM Channel definitions for the character device
+#define CDEV_PWM_REG_BASE (PWM_REG_BASE + PWM_NUM_CHANNELS * sizeof(u32))
+#define CDEV_PWM_REG_SIZE (sizeof(u32))
+#define CDEV_MAX_USERDATA 100
+
+#define PWM_CONVERSION_FACTOR (u32)(PWM_MAX / 100)
+#define TO_PWM_VALUE(x) ((x)*PWM_CONVERSION_FACTOR)
+#define FROM_PWM_VALUE(x) (u8)((x) / PWM_CONVERSION_FACTOR)
```

Comment:
eventuell besser in den Umrechenmakros vorher zu multiplizieren um Genauigkeitsverlust zu vermeiden.

#define TO_PWM_VALUE(x) (((x) * PWM_MAX)/100)